Index: app/src/main/java/com/example/fourthtask/ui/main/MainViewModel.kt
===================================================================
diff --git a/app/src/main/java/com/example/fourthtask/ui/main/MainViewModel.kt b/app/src/main/java/com/example/fourthtask/ui/main/MainViewModel.kt
new file mode 100644
--- /dev/null	(date 1768804472003)
+++ b/app/src/main/java/com/example/fourthtask/ui/main/MainViewModel.kt	(date 1768804472003)
@@ -0,0 +1,27 @@
+package com.example.fourthtask.ui.main
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.fourthtask.data.db.PostEntity
+import com.example.fourthtask.data.repository.PostRepository
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class MainViewModel(private val repository: PostRepository) : ViewModel() {
+
+    val postsLiveData = MutableLiveData<List<PostEntity>>()
+
+    fun loadPosts(isOnline: Boolean) {
+        viewModelScope.launch(Dispatchers.IO) {
+
+            if (isOnline) {
+                val apiPosts = repository.fetchPostsOnline()
+                repository.savePostsOffline(apiPosts)
+            }
+
+            val offlinePosts = repository.getOfflinePosts()
+            postsLiveData.postValue(offlinePosts)
+        }
+    }
+}
Index: app/src/main/java/com/example/fourthtask/data/repository/PostRepository.kt
===================================================================
diff --git a/app/src/main/java/com/example/fourthtask/data/repository/PostRepository.kt b/app/src/main/java/com/example/fourthtask/data/repository/PostRepository.kt
new file mode 100644
--- /dev/null	(date 1768804276500)
+++ b/app/src/main/java/com/example/fourthtask/data/repository/PostRepository.kt	(date 1768804276500)
@@ -0,0 +1,27 @@
+package com.example.fourthtask.data.repository
+
+import com.example.fourthtask.data.api.ApiService
+import com.example.fourthtask.data.db.PostDao
+import com.example.fourthtask.data.db.PostEntity
+import com.example.fourthtask.data.model.Post
+
+class PostRepository(
+    private val api: ApiService,
+    private val dao: PostDao
+) {
+
+    suspend fun fetchPostsOnline(): List<Post> {
+        return api.getPosts()
+    }
+
+    suspend fun savePostsOffline(posts: List<Post>) {
+        val entities = posts.map {
+            PostEntity(it.id, it.title, it.body)
+        }
+        dao.insertPosts(entities)
+    }
+
+    suspend fun getOfflinePosts(): List<PostEntity> {
+        return dao.getPosts()
+    }
+}
